pkg/agent/controller/bgp/controller.go

```diff
@@ -1009,8 +1065,8 @@ func (c *Controller) GetBGPPeerStatus(ctx context.Context, ipv4Peers, ipv6Peers
 }
 
 // GetBGPRoutes returns the advertised BGP routes.
-func (c *Controller) GetBGPRoutes(ctx context.Context, ipv4Routes, ipv6Routes bool) ([]string, error) {
-	getBgpRoutesAdvertised := func() sets.Set[bgp.Route] {
+func (c *Controller) GetBGPRoutes(ctx context.Context, ipv4Routes, ipv6Routes bool) (map[bgp.Route]RouteMetadata, error) {
+	getBgpRoutesAdvertised := func() map[bgp.Route]RouteMetadata {
 		c.bgpPolicyStateMutex.RLock()
 		defer c.bgpPolicyStateMutex.RUnlock()
```

Comment:
Not related to this PR. Just curious, should we got the read lock during copying `c.bgpPolicyState.routes`? Not just during getting it.

Reply:
I think you are correct as we don't get a copy of `c.bgpPolicyState.routes`. 
We can hold the lock for entire function `GetBGPRoutes` but if it is too long, we can also consider returning a copy of it during getting it ?

Reply:
I would say let's remove `getBgpRoutesAdvertised` and hold the lock for the whole function. The function itself is just basically making a copy anyway (with filter).

However, this is a bug, so please fix it in a separate PR that we can merge first and backport to Antrea v2.2

Thanks for catching this @hongliangl